### PR TITLE
refactor(lambda): move log retention custom resource to logs (#9671)

### LINKED ISSUE #9671: [lambda][logs] Make LogRetention construct reusable for any log group not just Lambda's
Various AWS resources autocreate log groups if they don't exist e.g. Lambda functions, SageMaker Endpoints.

For a while, my team have managed to create log groups explicitly for lambda functions to customize the log retention policies. This worked well because Lambda didn't attempt to create the log groups until the first request. However, I believe that with recent changes over the past year in Lambda (VPC improvements, Provisioned Concurrency), log groups can be created before the first request which makes our approach slightly problematic.

We noticed the introduction of [LogRetention construct](https://github.com/aws/aws-cdk/blob/014c13a78261b400404819549f6ff25d27b0c51d/packages/%40aws-cdk/aws-lambda/lib/log-retention.ts) and log retention props for Lambda functions that utilize it to let the function create the log group and have a custom resource that manages setting the retention policy.

We think that the LogRetention construct can be reused outside of just Lambda functions e.g. SageMaker Endpoints.

This is a feature request to update LogsRetention construct and the [provider](https://github.com/aws/aws-cdk/blob/014c13a78261b400404819549f6ff25d27b0c51d/packages/%40aws-cdk/aws-lambda/lib/log-retention-provider/index.ts) it uses to make them reusable for non-Lambda function log groups.

### Use Case

Enable using the LogRetention construct for any AWS resource that auocreates logs to manage those logs retention.
Out of scope is introducing different props to different resource constructs to add that support. That can come in later changes.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

As my team might be able to contribute to this work, we're asking for feedback on this approach. If you think it's sensible, then one of the questions we have is about the location of this construct (it's currently part of the lambda module) but if it's to become general purpose, it probably should go somewhere else e.g. the logs module.

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by humanzz**
move LogRetention construct definition from lambda to logs while refactoring it so it does not depend on lambda constructs
this required reimplementing the functionality provided by lambda.SingletonFunction using CfnResource

keep declared classes/interfaces in lambda for compatability while marking them as deprecated
they should be removed in an upcoming breaking change for their current customers in lambda and rds

remove rds dependency on lambda

Fixes #9671

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by nija-at — CHANGES_REQUESTED** [review]
Nice! I like how this is coming out. 

Just one comment about the logical ids below. It feels like we can go all the way and have no diffs in the generated stacks.

File: packages/@aws-cdk/aws-lambda/test/integ.log-retention.expected.json [hunk 1]
```diff
@@ -73,7 +73,7 @@
         "RetentionInDays": 7
       }
     },
-    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
```
[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-lambda/test/integ.log-retention.expected.json:76]: Avoiding the logical Id changes is a great idea.
Does the construct type play no role at all in the generated Id? Just the hierarchy/path?

[the remark this answers is not in the document]
Comment (nija-at) [reply packages/@aws-cdk/aws-lambda/test/integ.log-retention.expected.json:76]: I believe only the hierarchy and the names in the hierarchy play a role in how this id is generated.

I'm not intimately familiar with this area of the code base but the logic resides here - 

https://github.com/aws/aws-cdk/blob/b1d0ac0564a86ab325e06b18670657ee9c953e3e/packages/%40aws-cdk/core/lib/stack.ts#L733-L778

[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-lambda/test/integ.log-retention.expected.json]: So a new revision is up doing just that... the logical ids for the role do not change... however, the hash part of the function resource is now dropped.

The only difference in tree structure i can think of is that
- Previously: `LogRetention` -> `SingletonFunction` -> `LambdaFunction` (but `LambdaFunction` used stack as its parent despite being created inside `SingletonFunction`)
- The new revision: `LogRetention` -> `LogRetentionFunction` where `LogRetentionFunction` uses stack as its parent abd being created inside `LogRetention`.

Strangely, the children of `LogRetentionFunction` get the same logical ids as before, but itself doesn't.

[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-lambda/test/integ.log-retention.expected.json]: @nija-at on a further look, i actually think, based on https://github.com/aws/aws-cdk/blob/b1d0ac0564a86ab325e06b18670657ee9c953e3e/packages/%40aws-cdk/core/lib/stack.ts#L755, that the issue now is LambdaFunction itself is a construct… the actual cfnresource is a child of it… that’s why the cfn resource gets a hash
In the new implementation… LogRetentionFunction is itself a cfnresource… so when added to the top level it doesn’t get a hash
Any suggestions to get around that?

[the remark this answers is not in the document]
Comment (nija-at) [reply packages/@aws-cdk/aws-lambda/test/integ.log-retention.expected.json]: How about artificially introducing a `Construct` in the new hierarchy to reproduce the original hierarchy. Have you tried that?

**Comment by humanzz** [comment]
Apologies for the multiple noisy commits, my local environment is having issues and I'm using CodeBuild for build verification. Changed the PR to draft mode till I get a proper working version.

File: packages/@aws-cdk/aws-lambda/test/integ.log-retention.expected.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-lambda/test/integ.log-retention.expected.json]: yup... that worked... this change is now ready for review

File: packages/@aws-cdk/aws-lambda/lib/log-retention.ts [hunk 2]
```diff
@@ -1,116 +1,31 @@
...
-   */
-  readonly role?: iam.IRole;
-
-  /**
-   * Retry options for all AWS API calls.
-   *
-   * @default - AWS SDK default retry options
-   */
-  readonly logRetentionRetryOptions?: LogRetentionRetryOptions;
+export interface LogRetentionRetryOptions extends logs.LogRetentionRetryOptions {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda/lib/log-retention.ts:9]: This isn't used anywhere now, but I assume it technically needs to exist for backwards compatibility, yes?

File: packages/@aws-cdk/aws-logs/test/test.log-retention.ts [hunk 3]
```diff
@@ -1,8 +1,8 @@
 import { ABSENT, countResources, expect, haveResource } from '@aws-cdk/assert';
 import * as iam from '@aws-cdk/aws-iam';
-import * as logs from '@aws-cdk/aws-logs';
 import * as cdk from '@aws-cdk/core';
 import { Test } from 'nodeunit';
+import { RetentionDays } from '../lib';
 import { LogRetention } from '../lib/log-retention';
```
Comment (nija-at) [inline packages/@aws-cdk/aws-logs/test/test.log-retention.ts:6]: Minor

```suggestion
import { LogRetention, RetentionDays } from '../lib';
```

File: packages/@aws-cdk/aws-logs/lib/log-retention.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+  /**
+   * Helper method to ensure that only one instance of LogRetentionFunction resources are in the stack mimicking the
+   * behaviour of @aws-cdk/aws-lambda's SingletonFunction to prevent circular dependencies
+   */
+  private ensureSingletonLogRetentionFunction(props: LogRetentionProps) {
+    const functionLogicalId = 'LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8a';
+    const existing = cdk.Stack.of(this).node.tryFindChild(functionLogicalId);
+    if (existing) {
+      // Just assume this is true
```
Comment (nija-at) [inline packages/@aws-cdk/aws-logs/lib/log-retention.ts:107]: Delete this comment. Not clear what this means.

```suggestion
```

File: packages/@aws-cdk/aws-rds/package.json [hunk 5]
```diff
@@ -64,6 +64,7 @@
   "devDependencies": {
     "@aws-cdk/assert": "0.0.0",
     "@aws-cdk/aws-events-targets": "0.0.0",
+    "@aws-cdk/aws-lambda": "0.0.0",
```
Comment (nija-at) [inline packages/@aws-cdk/aws-rds/package.json:67]: Why is this needed as a dev dependency?

**Review by nija-at — CHANGES_REQUESTED** [review]
Good work on getting the logical ids aligned. Just a couple of comments around union types and dependencies.

File: packages/@aws-cdk/aws-lambda/lib/log-retention.ts (hunk 2 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-lambda/lib/log-retention.ts:9]: Yeah... If anyone uses it in their own code

File: packages/@aws-cdk/aws-rds/package.json (hunk 5 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-rds/package.json:67]: for testing purposes as there are a couple of non-logretention-related tests that use lambda e.g.

https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-rds/test/test.instance.ts#L340

https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-rds/test/integ.instance.lit.ts#L85

File: packages/@aws-cdk/aws-logs/lib/log-retention.ts (hunk 4 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-logs/lib/log-retention.ts:107]: the effect of copy/pasting existing code

File: packages/@aws-cdk/aws-lambda/lib/function.ts [hunk 6]
```diff
@@ -256,7 +256,7 @@ export interface FunctionOptions extends EventInvokeConfigOptions {
    *
    * @default - Default AWS SDK retry options.
    */
-  readonly logRetentionRetryOptions?: LogRetentionRetryOptions;
+  readonly logRetentionRetryOptions?: LogRetentionRetryOptions | logs.LogRetentionRetryOptions;
```
[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-lambda/lib/function.ts]: done

File: packages/@aws-cdk/aws-logs/test/test.log-retention.ts (hunk 3 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-logs/test/test.log-retention.ts]: done

**Review by nija-at — APPROVED** [review]
All clear 🙌

**Comment by humanzz** [comment]
Great... looking forwards to this and the other changes we have lined up.

**Comment by humanzz** [comment]
I'm merging and resolving conflicts

**Review by nija-at — APPROVED** [review]

**Comment by humanzz** [comment]
@nija-at seems like automerge didn’t kick off for some reason. It’s not clear to why that didn’t happen so appreciate letting me know if any further changes are needed on my end

**Comment by nija-at** [comment]
@Mergifyio refresh

**Comment by nija-at** [comment]
@humanzz - not sure what's going on with mergify there. I've manually merged this PR in for you.

**Comment by humanzz** [comment]
@nija-at a little hiccup it seems but glad we got this merged. Thanks :-)
